src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -81,6 +81,64 @@ internal BufferingJsonReader(IJsonReader innerReader, string inStreamErrorProper
             this.currentBufferedNode = null;
         }
 
+        internal bool TryGetValueFromBuffered(string propertyName, bool removeIfExist, out object value)

```

Comment:
This could have some pretty severe performance ramifications.

If a payload doesn't have the value that we try to get, we will read the entire payload into memory.  For very large, or streamed, responses, this could overflow memory.

This kinda defeats the whole point of streaming and ordering, which is that we can process the response in order.

The previous implementation assumes that type, if present, is the first thing we read. So we read the first node and, if it's type, we store that, and if it is not type, we assume there is no type and continue reading the payload as normal.

I would have expected we simply expand that logic.  That is, we read the first node. If it is one of our "metadata" nodes (context, removed, type, id), then we get the value and repeat for the next node. As soon as we find a node that is not one of our metadata nodes, we break and start reading the payload as normal.